specs/ThrottlingControlScript.md

```diff
@@ -0,0 +1,291 @@
+[uuid(00f1b5fb-91ed-4722-9404-e0f8fd1e6b0a), object, pointer_default(unique)]
+interface ICoreWebView2_21 : ICoreWebView2_20 {
+  /// Get the preferred wake up interval (in milliseconds) for throttleable
+  /// JavaScript tasks (`setTimeout` and `setInterval`), for the given
+  /// throttling category. A wake up interval is the amount of time that needs
+  /// to pass before the WebView2 Runtime checks for new timer tasks to run.
+  /// The default interval values are constants determined by the running
+  /// version of the WebView2 Runtime.
+  HRESULT GetThrottlingIntervalPreference(
+      [in] COREWEBVIEW2_THROTTLING_CATEGORY category,
+      [out, retval] UINT32* intervalInMilliseconds);
+
+  /// Sets the preferred wake up interval (in milliseconds) for throttleable
+  /// JavaScript tasks (`setTimeout` and `setInterval`), for the given
+  /// throttling category. A wake up interval is the amount of time that needs
+  /// to pass before the WebView2 Runtime checks for new timer tasks to run. For
+  /// example, an application might use a foreground value of 30 ms for moderate
+  /// throttling scenarios, or match the default background value (usually 1000
+  /// ms). The WebView2 Runtime will try to respect the preferred interval set
+  /// by the application, but the effective value will be constrained by
+  /// resource and platform limitations. Setting a value of `0` means no
+  /// throttling will be applied.
+  HRESULT SetThrottlingIntervalPreference(
+      [in] COREWEBVIEW2_THROTTLING_CATEGORY category,
+      [in] UINT32 intervalInMilliseconds);
+}
+
+/// A continuation of the `ICoreWebView2Frame` interface to support IsUntrusted property.
+[uuid(5b7d1b96-699b-44a2-b9f1-b8e88f9ac2be), object, pointer_default(unique)]
+interface ICoreWebView2Frame6 : ICoreWebView2Frame5 {
+  /// The `IsUntrusted` property indicates whether the frame has been marked
+  /// untrusted by the host app. Untrusted frames will receive a different
+  /// script throttling category as compared to regular frames. Defaults to
+  /// `FALSE` unless set otherwise. When `FALSE`, and for main frame, throttling
+  /// category will be determined by page state. The corresponding preferred
+  /// interval will apply (set through `SetThrottlingIntervalPreference`).
+  [propget] HRESULT IsUntrusted([out, retval] BOOL* value);
+
+  /// Marks the frame as untrusted, for script throttling purposes.
+  [propput] HRESULT IsUntrusted([in] BOOL value);
+}
+
+```
+
+```C#
+namespace Microsoft.Web.WebView2.Core
+{
+    enum CoreWebView2ThrottlingCategory
+    {
+        Foreground = 0,
+        Background = 1,
+        Intensive = 2,
+        UntrustedFrame = 3,
+    };
+
+    runtimeclass CoreWebView2
+    {
+        // ...
+
+        [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2_21")]
+        {
+            UInt32 GetThrottlingIntervalPreference(CoreWebView2ThrottlingCategory category);
```

Comment:
I'm wondering if instead of two methods on the CoreWebView2, we should make this four get/set properties on the CoreWebView2Settings.

Reading through your document it doesn't seem like its useful to have a method that takes this as an enum because the caller will always be explicitly changing the value of a specific category. In which case its probably easier to have individual properties for each category rather than a method where you choose the category.

And then the whole scenario seems very uncommon so rather than put this on the CoreWebView2 we can move it at least to the CoreWebView2Settings.

What do you think?

Reply:
I have no strong opinion on whether to put these in `CoreWebView2` or `CoreWebView2Settings`. Looking through our existing APIs, I see that there are properties in the settings object which are applied immediately, so I think we can use the same pattern here.

For enum vs methods. I don't have a strong preference either, but I'm not sure I understand the reasoning:

> the caller will always be explicitly changing the value of a specific category

isn't this what we do for other APIs taking non-flags enum types?
Only thing is I think the enum makes it easier to find what the possible/configurable states are.

Reply:
So this is what it would look like moving as methods to `CoreWebView2Settings`. Is this what you are suggesting?

```
namespace Microsoft.Web.WebView2.Core
{

    runtimeclass CoreWebView2Settings
    {
        // ...

        UInt32 ForegroundThrottlingIntervalPreference { get; set; }
        UInt32 BackgroundThrottlingIntervalPreference { get; set; }
        UInt32 IntensiveThrottlingIntervalPreference { get; set; }
        UInt32 IsolatedThrottlingIntervalPreference { get; set; }
    }

    runtimeclass CoreWebView2Frame
    {
        // ...
        
        Boolean ShouldUseIsolatedThrottling { get; set; };
    }
}

```

Reply:
Undecided on the name pattern though. I think we'd want them to show up together so maybe the pattern should be something like `ThrottlingIntervalPreference<category>` instead?

Reply:
Giving them the same prefix does help group them although the name is less readable. Let's try that in the API review and see what they say

Reply:
Just pushed the update moving to properties in settings object. Thanks!